Fix array handling, fix tests, other smaller fixes

GUI opened the pull request:
I have a number of commits here to fix a few issues I encountered and also resurrect the test suite for this project. Sorry to pile all this into a single pull request, but hopefully each commit is pretty explanatory in what it fixes. I could try to pull them apart into separate pull requests, but they all sort of become part of a larger effort to get the entire test suite passing again (I had started with just wanting to fix issues with arrays, but I wanted to get some sort of verification I wasn't breaking other things, so that led me down this testing rabbit-hole).

In any case, here's a summary of what's changed and fixed:
- Fixes wonkiness with array storage and fetching, related to this issue: https://github.com/bigplum/lua-resty-mongol/issues/29 (I'd consider the previous behavior broken, but see my comment for more detail the behavior change: https://github.com/GUI/resty-mongol/commit/9c6d2c5b8d099670c0f5e0cd87d77d0305e05e43)
- Fixes PID handling in OS X: https://github.com/bigplum/lua-resty-mongol/pull/25 Also improves efficiency and allows the module to be used inside `init_worker_by_lua` by using `ngx.worker.pid()`.
- Fixes issues with `cursor:limit()` that the existing tests uncovered (in some cases, the limit wasn't being applied and the cursor would iterate over all the results, despite having a limit).
- Makes some minor fixes in edge-cases around gridfs and authentication that the existing tests uncovered.
- I synced up the test suite with the more recent additions to [bigplum/lua-resty-mongol](https://github.com/bigplum/lua-resty-mongol/commits/master).
- I added a `make test` command that should do everything necessary to run the tests.
- I added configuration for CircleCI, so you can have the tests automatically run there, where all tests are now passing: https://circleci.com/gh/GUI/resty-mongol/9
- Updated the text in the README to take into account this module's different name (`resty-mongol` instead of `resty.mongol`) and it's different style of constructor for the connection.

Let me know if you have any questions/comments/concerns about any of this. Thanks!

GUI committed:
Add an automated one-command way to run the tests.

This installs the necessary dependencies (openresty and Test::Nginx) and
starts up and tears down the mongodb replicaset needed for tests.

GUI committed:
Sync up tests with latest from upstream bigplum/lua-resty-mongol

GUI committed:
Update tests to account for the renamed module on this fork.

GUI committed:
Fix authentication tests. Make authentication failures non-fatal.

- If authentication failed, the app was raising 500 error, rather than
  returning an error.
- The authentication failure messages coming from MongoDB appear to have
  changed under MongoDB 2.6. Fix these tests by turning them into
  regexes so we can pass regardless of MongoDB being tested against.

GUI committed:
Fix gridfs remove errors when trying to remove nonexistent results.

This fixes TEST 3 and TEST 4 in t/gridfs.t since they attempt to delete
nonexistent results. MongoDB was erroring since the "$in" clause didn't
like being passed an empty array.

GUI committed:
Fix gridfs file not being persisted with the correct file_size.

This fixes TEST 14 and TEST 15 in t/gridfs_write.t, since those tests
were using :write to add data, but then the file_size wasn't being set,
so the :read was then failing (since it thought the file should be
empty).

GUI committed:
Fix gridfs_write.t test failing due to misnumbered tests.

Test::Nginx thought 59 tests should have been run instead of the 45
actual tests that exist.

GUI committed:
Fix array handling.

Previously if you fetched an record from MongoDB that contained an
array, it would get translated incorrectly into a Lua table with a [0]
index (rather than starting at 1). This would cause problems when trying
to iterate over arrays fetched from MongoDB.

Furthermore, when this adapter was used to insert arrays, the record
actually inserted in MongoDB was a hash object instead of an array. This
adapter would parse that type of object back as an array in Lua, so
that's why the current tests around arrays were passing. So in order to
properly test this, we'll use the mongo command line client to insert
some structures directly, and then make sure that real MongoDB arrays
get read back out as Lua array-type tables.

This also fixes broken tests around the replicaset getprimary()
function, since those were failing since MongoDB was returning an
array, which wasn't being iterated over properly.

This change does mean that an explicit [0] index type of table in Lua
has to change behavior, since we can't store that as a real Mongo array
and also know that the index should start at 0 in Lua. So these type of
arrays get converted to normal arrays upon writes, and then will get
read as an array (where the index will start with Lua's normal [1]
index). I'm not sure that's the best option (the "cjson" library
converts these types of [0] index arrays to hash-like objects upon
serialization instead of arrays), but it seems like this is the simplest
way to address these edge-case arrays for now.

See:
https://github.com/bigplum/lua-resty-mongol/issues/29
https://github.com/ashun/lua-resty-mongol/commit/9c4f9aaa223eb63f2505c24a2df150b1975b9f2f

GUI committed:
Improve PID handling and fix PID problems in OS X.

- Switch from "ngx.var.pid" to "ngx.worker.pid()". This lets the module
  be used in "init_worker_by_lua" and is also supposed to be more
  efficient: http://wiki.nginx.org/HttpLuaModule#ngx.worker.pid
- Fix random failures in OS X if the worker PIDs happen to be higher
  than expected numbers:
  https://github.com/bigplum/lua-resty-mongol/pull/25

GUI committed:
Fix cursor limits not always being applied.

This fixes TEST 1 in cursor.t, where the cursor was fetching more
results than the limit specified.

This also fixes some other tests that were apparently relying on the
"num_each_query" parameter to the col:find() calls applying a query
limit. This correctly no longer applies a limit (since that parameter
just controls how many results are in each cursor fetch), so explicit
limits needed to be added to these few test places.

GUI committed:
Fix cursor:sort() tests. All tests should be passing (fingers crossed)

Since this fork performs sorting server-side, the behavior is different
than the in-lua sorting that was happening before.

GUI committed:
Fix readme references to old module name and connection instructions.

GUI committed:
Adding Circle CI setup for testing.

We have to install MongoDB 2.6 since some of the GridFS functionality
uses the update_all command which uses the new "update" db command
that's only available in MongoDB 2.6.

DorianGray commented:
This looks super solid. Thanks so much! I'll review a little closer, make a few additions, and merge shortly.

DorianGray commented:
Funny, because this was literally my next goal.

DorianGray commented:
@GUI I've tested out these changes. They look good, with the exception of them breaking skip. I've marked what I believe is the offending line, would you mind looking at a different way to solve that particular problem? Skip was never implemented properly in bigplum's version, I implemented it here but it did have some issues(that were able to be worked around).

DorianGray commented on src/cursor.lua:
```diff
@@ -59,7 +59,7 @@ function cursor_methods:count()
 
 function cursor_methods:next()
-  if self.limit_n > 0 and self.i - self.skip_n >= self.limit_n then return nil end
+  if self.limit_n > 0 and self.i >= self.limit_n then return nil end
```
I believe this breaks skip.

DorianGray replied:
To be explicit, we don't want to keep -all- records in memory, we only want to keep this set of records in memory. We return a generator, not a list.

GUI committed:
Fix STDERR redirection syntax for pre-4.0 Bash

GUI committed:
Fix skip() in a way that also fixes limit().

The earlier fix (8bec0f296054f6517d374589e74835f3f3c8198b) for limits on
queries where num_each_query was specified broke skip. This approach
should fix both scenarios.

GUI committed:
Fix invalid query in TEST 1 of t/objid.t.

This started consistently breaking after the additional tests added in
f1575798dc99b2dc4d59cb90804dfc7db5cd3121. This test query wasn't
actually performing a conditional query, since it was querying on a nil
value (r["id"] rather than r["_id"]). So it was just sheer luck that the
test somehow used to fetch the correct record and pass. But the
additional records inserted as part of the new tests managed to unearth
this issue with the test.

GUI commented:
I believe `skip` is now fixed in a way that also fixes the original `limit` batching issue with the changes in: https://github.com/GUI/resty-mongol/commit/f1575798dc99b2dc4d59cb90804dfc7db5cd3121 I also added tests around the skip function. Did you want to take another look to see if this makes sense and appears correct to you? To fix this, I added an `overall_index` variable to keep track of where the cursor is in the complete result set (regardless of batches), whereas the `i` variable now just keeps track of where the cursor is inside the single batch of results that's currently in memory.

I also uncovered an issue with the `t/objid.t` test and fixed that in https://github.com/GUI/resty-mongol/commit/553cde4e3ceb9390fcb1ba72d4bae933802d5008.

johnae commented:
@DorianGray @GUI would be great if this was released. That mongo:new() vs mongo() really had me confused for a long time - thanks for documenting this somewhat different usage.
